File: model/features.py
"""Pure feature engineering for xPts v1. No I/O; operates on DataFrames."""
from __future__ import annotations

from typing import Sequence

import numpy as np
import pandas as pd

from feature_spec import (
    DECAY_ALPHA,
    FORM_STATS,
    FORM_WINDOW,
    STRENGTH_SCALE,
    VALUE_SCALE,
    FEATURE_COLUMNS,
)


def decay_weights(n: int, alpha: float = DECAY_ALPHA) -> np.ndarray:
    if n <= 0:
        return np.array([])
    w = alpha ** np.arange(n)
    return w / w.sum()


def exp_decay_mean(values_recent_first: Sequence[float], alpha: float = DECAY_ALPHA) -> float:
    vals = list(values_recent_first)
    if not vals:
        return 0.0
    w = decay_weights(len(vals), alpha)
    return float(np.dot(w, np.asarray(vals, dtype=float)))


def opponent_strengths(was_home: bool, opponent_team: int,
                       team_strengths: dict[int, dict]) -> tuple[float, float]:
    t = team_strengths.get(int(opponent_team))
    if t is None:
        return (0.0, 0.0)
    # If the player is home, the opponent plays away (use opponent's away strengths).
    if was_home:
        d, a = t["strength_defence_away"], t["strength_attack_away"]
    else:
        d, a = t["strength_defence_home"], t["strength_attack_home"]
    return (float(d) / STRENGTH_SCALE, float(a) / STRENGTH_SCALE)


def build_feature_row(prior_rows: pd.DataFrame, target_row: pd.Series,
                      team_strengths: dict[int, dict]) -> dict:
    # Most-recent-first, capped at the form window.
    prior = prior_rows.sort_values(["gw", "fixture_id"], ascending=False).head(FORM_WINDOW)

    feat: dict[str, float] = {}
    for stat in FORM_STATS:
        feat[f"form_{stat}"] = exp_decay_mean(prior[stat].tolist())

    # xMinutes proxy: recent starts share (0..1). 0 if no prior rows.
    feat["xmin"] = float(prior["starts"].mean()) if len(prior) else 0.0

    opp_def, opp_att = opponent_strengths(
        bool(target_row["was_home"]), int(target_row["opponent_team"]), team_strengths
    )
    feat["opp_strength_def"] = opp_def
    feat["opp_strength_att"] = opp_att
    feat["was_home"] = 1.0 if bool(target_row["was_home"]) else 0.0
    feat["value_scaled"] = float(target_row["value"]) / VALUE_SCALE
    return feat


def build_samples(history: pd.DataFrame, team_strengths: dict[int, dict]) -> pd.DataFrame:
    rows = []
    for player_id, pdf in history.groupby("player_id"):
        pdf = pdf.sort_values(["gw", "fixture_id"])
        for i in range(len(pdf)):
            target = pdf.iloc[i]
            prior = pdf[pdf["gw"] < target["gw"]]
            if len(prior) == 0:
                continue  # need at least one prior gameweek for features
            feat = build_feature_row(prior, target, team_strengths)
            feat.update({
                "player_id": int(player_id),
                "gw": int(target["gw"]),
                "position": target["position"],
                "target": float(target["total_points"]),
                "actual_minutes": int(target["minutes"]),
            })
            rows.append(feat)
    cols = FEATURE_COLUMNS + ["player_id", "gw", "position", "target", "actual_minutes"]
    return pd.DataFrame(rows, columns=cols)

File: supabase/functions/fpl-project/lib/features.ts
import { DECAY_ALPHA, FORM_STATS, FORM_WINDOW, STRENGTH_SCALE, VALUE_SCALE } from '../feature-spec.ts';

export interface HistoryRow {
  gw: number;
  fixture_id: number;
  starts: number;
  expected_goals: number;
  expected_assists: number;
  expected_goal_involvements: number;
  threat: number;
  creativity: number;
  influence: number;
  bps: number;
  defensive_contribution: number;
  total_points: number;
}

export interface ClubStrength {
  strength_defence_home: number;
  strength_defence_away: number;
  strength_attack_home: number;
  strength_attack_away: number;
}

export interface FixtureTarget {
  was_home: boolean;
  opponent_team: number;
  value: number;
}

export function decayWeights(n: number, alpha: number = DECAY_ALPHA): number[] {
  if (n <= 0) return [];
  const w = Array.from({ length: n }, (_, i) => alpha ** i);
  const sum = w.reduce((a: number, b: number) => a + b, 0);
  return w.map((x) => x / sum);
}

export function expDecayMean(valuesRecentFirst: number[], alpha: number = DECAY_ALPHA): number {
  if (valuesRecentFirst.length === 0) return 0;
  const w = decayWeights(valuesRecentFirst.length, alpha);
  let total = 0;
  for (let i = 0; i < w.length; i++) total += w[i] * valuesRecentFirst[i];
  return total;
}

export function opponentStrengths(
  wasHome: boolean,
  opponentTeam: number,
  clubStrengths: Record<number, ClubStrength>,
): { def: number; att: number } {
  const t = clubStrengths[opponentTeam];
  if (!t) return { def: 0, att: 0 };
  // player home -> opponent plays away -> use opponent's away strengths
  const def = wasHome ? t.strength_defence_away : t.strength_defence_home;
  const att = wasHome ? t.strength_attack_away : t.strength_attack_home;
  return { def: def / STRENGTH_SCALE, att: att / STRENGTH_SCALE };
}

export function buildFeatureRow(
  priorRows: HistoryRow[],
  target: FixtureTarget,
  clubStrengths: Record<number, ClubStrength>,
): Record<string, number> {
  // most-recent-first by (gw, fixture_id), capped at the form window
  const prior = [...priorRows]
    .sort((a, b) => (b.gw - a.gw) || (b.fixture_id - a.fixture_id))
    .slice(0, FORM_WINDOW);

  const feat: Record<string, number> = {};
  for (const stat of FORM_STATS) {
    feat[`form_${stat}`] = expDecayMean(prior.map((r) => Number(r[stat as keyof HistoryRow])));
  }
  feat.xmin = prior.length ? prior.reduce((a: number, r: HistoryRow) => a + Number(r.starts), 0) / prior.length : 0;

  const opp = opponentStrengths(target.was_home, target.opponent_team, clubStrengths);
  feat.opp_strength_def = opp.def;
  feat.opp_strength_att = opp.att;
  feat.was_home = target.was_home ? 1 : 0;
  feat.value_scaled = Number(target.value) / VALUE_SCALE;
  return feat;
}
